File: contracts/dre-contract/src/actions/insert.ts
import { ContractState, ContractInput, ContractResult } from '../types'

export default function insert(state: ContractState, input: ContractInput): ContractResult {
    if (!input.contentType) throw new ContractError('invalid arguments')

    const caller = SmartWeave.transaction.owner
    
    if (!state.hierarchy.files[caller]) state.hierarchy.files[caller] = {
        [caller]: {
            id: caller,
            name: caller,
            contentType: 'folder',
            children: []
        }
    }

    if (!input.id) input.id = SmartWeave.transaction.id
    if (!input.parentId) input.parentId = caller

    var files = state.hierarchy.files[caller]

    if (files[input.id]) throw new ContractError('file already exists')    

    files[input.id] = {
        id: input.id,
        parentId: input.parentId,
        contentType: input.contentType,
        children: [],
        name: input.name,
        size: input.size
    }

    const parent = files[input.parentId] 

    if (!parent) throw new ContractError('file parent does not exist')

    if (parent.contentType != 'folder') throw new ContractError('parent is not a folder')

    parent.children.push(input.id)

    return { state }
}

File: contracts/client-contract/src/actions/insert.rs
use crate::types::{ ContractState, Insert, File };
use warp_contracts::{ js_imports::Transaction, handler_result::WriteResult };

pub fn insert(mut state: ContractState, input: Insert) -> WriteResult<ContractState, String> {
    if !state.is_owner() {
        return WriteResult::ContractError("invalid owner".to_string())
    }
    
    let id = input.id.unwrap_or(Transaction::id());

    if state.hierarchy.contains(&id) {
        return WriteResult::ContractError("file is already in the hierarchy".to_string())
    }

    state.hierarchy.insert_file(File {
        id: id.clone(),
        name: input.name,
        parent_id: input.parent_id.clone(),
        prev_parent_id: None,
        size: input.size,
        children: if input.content_type.eq("folder") { Some(Vec::new()) } else { None },
        created_at: input.created_at,
        content_type: input.content_type,
        encryption: input.encryption,
    });

    let parent_id = input.parent_id.unwrap_or("root".to_string());
    let parent = state.hierarchy.get_file_mut(&parent_id);

    if parent.is_none() || parent.as_ref().unwrap().children.is_none() {
        return WriteResult::ContractError("parent does not exist".to_string())
    }

    parent.unwrap().add_child(id);

    WriteResult::Success(state)
}